Added SADD, SREM, SMEMBERS, SCARD, SINTER, SDIFF

apoorvyadav1111 opened the pull request:
### The changes that are included in the PR.
Added support for the set data structure and related commands as per Redis.
The following commands have been added and further more will be added in future PRs. 

1. SADD
2. SREM
3. SMEMBERS
4. SCARD
5. SDIFF
6. SINTER



### Information on any implementation choices that were made.

I have created a set structure using map since go does not support set as a builtin type. For object encoding, dedicated types have been added for checking object type while mutating the set using commands. 

For the commands and their optimization, I have attempted to follow the Redis protocol for these commands in scenarios such as:
1. error handling and cases for all scenarios like non-existing key, and invalid type of the object value
2. touching the keys even if there is an early return due to optimization incase of SDIFF and SINTER.
3. skipping the touch for subsequent keys if we access an object where the type is not SET.
4. optimizing the SDIFF and SINTER as per Redis implementation. 


Evidence of sufficient testing. You MUST indicate the tests done, either manually or automated.
Added new tests for all the scenarios in the tests folder

I have added only six commands as the size of the PR would have been large for just one merge and review. If the implementation looks good, I can pick up other complex commands as well. 

If a design document is required, I can create the same. I wanted to get the changes reviewed before spending more time.

apoorvyadav1111 commented:
Hi @JyotinderSingh, Please review the changes/ assign a reviewer at your convenience.

Thanks,
Apoorv

JyotinderSingh commented on core/t_set.go:
```diff
@@ -0,0 +1,290 @@
+
+	// Get the set object.
+	set := obj.Value.(map[string]bool)
+	for i := 1; i < len(args); i++ {
+		if set[args[i]] {
+			delete(set, args[i])
+			count++
+		}
+	}
```
```suggestion
	for _, arg := range args[1:] {
	if set[arg] {
		delete(set, arg)
		count++
	}
}

```
same for all other such loops in this PR.

JyotinderSingh commented on core/t_set.go:
```diff
@@ -0,0 +1,290 @@
+	}
+
+	// Get the members of the set.
+	var members []string
```
```suggestion
	members := make([]string, 0, len(srcSet))
```

This improves memory allocation performance by avoiding reallocs while adding the elements in.

JyotinderSingh commented on core/t_set.go:
```diff
@@ -0,0 +1,290 @@
+		}
+	}
+
+	for i := 1; i < len(args); i++ {
```
```suggestion
	for _, key := range args[1:] {
```

JyotinderSingh commented on core/t_set.go:
```diff
@@ -0,0 +1,290 @@
+
+	var empty int = 0
+
+	for i := 0; i < len(args); i++ {
+		// Get the set object from the store.
+		obj := Get(args[i])
+
+		if obj == nil {
+			empty++
+			continue
+		}
+
+		// If the object exists, check if it is a set object.
+		if obj.TypeEncoding != ObjTypeSet|ObjEncodingHT {
+			return Encode(errors.New("WRONGTYPE Operation against a key holding the wrong kind of value"), false)
+		}
+
+		// Get the set object.
+		set := obj.Value.(map[string]bool)
+		sets[i] = set
+	}
```
```suggestion
	for _, key := range args {
		obj := Get(key)

		if obj == nil {
			// If any set is missing, return an empty result immediately
			return Encode([]string{}, false)
		}

		// Ensure the object is a set with the correct encoding
		if obj.TypeEncoding != ObjTypeSet || obj.Encoding != ObjEncodingHT {
			return Encode(errors.New("WRONGTYPE Operation against a key holding the wrong kind of value"), false)
		}

		// Add the set to the list of sets
		sets = append(sets, obj.Value.(map[string]bool))
	}
```

More idiomatic way to write this. Additionally you can directly return if you ever encounter a nil set instead of maintaining a count.

JyotinderSingh commented on core/t_set.go:
```diff
@@ -0,0 +1,290 @@
+	// init the result set with the first set
+	// store the number of elements in the first set in count
+	// we will decrement the count if we do not find the elements in the other sets
+	for k, flag := range sets[0] {
+		if !flag {
+			continue
+		}
+		count++
+		resultSet[k] = true
+	}
```
```suggestion
	for k, flag := range sets[0] {
		if flag {
			resultSet[k] = true
		}
	}
```

Better readability.

JyotinderSingh commented on core/t_set.go:
```diff
@@ -0,0 +1,290 @@
+		resultSet[k] = true
+	}
+
+	for i := 1; i < len(sets); i++ {
+		if count == 0 {
+			break
+		}
+		for k, flag := range resultSet {
+			if !flag {
+				continue
+			}
+			if !sets[i][k] {
+				delete(resultSet, k)
+				count--
+			}
+		}
+	}
```
```suggestion
	for i := 1; i < len(sets); i++ {
		for k := range resultSet {
			if !sets[i][k] {
				delete(resultSet, k)
			}
		}
	}

```

Improve readability.

JyotinderSingh commented on core/t_set.go:
```diff
@@ -0,0 +1,290 @@
+			}
+		}
+	}
+	for k, flag := range resultSet {
+		if flag {
+			members = append(members, k)
+		}
+	}
+	return Encode(members, false)
```
```suggestion
	members := make([]string, 0, len(resultSet))
	for k := range resultSet {
		members = append(members, k)
	}

	return Encode(members, false)
```
Improve readability.

JyotinderSingh reviewed (changes requested):
Thanks for these changes @apoorvyadav1111. I have left some comments.

JyotinderSingh commented on core/commands.go:
```diff
@@ -570,6 +570,61 @@ var (
 		Arity:    -2,
 		KeySpecs: KeySpecs{BeginIndex: 1, Step: 1},
 	}
+	saddCmdMeta = DiceCmdMeta{
```
Please add these new initializers to the end of the existing list.

AshwinKul28 commented:
Thanks @apoorvyadav1111 for the changes. Apart from the suggested changes by Jyotindar, please check the many errors from the lint. Maybe the branch is out of sync with the master. Just pull the latest once, many issues might get resolved automatically.

apoorvyadav1111 commented:
Hi @JyotinderSingh , I have addressed the comments from the PR and will do a rebase with the master and force-push the branch for further review (to fix lint and build errors). I see that we have a multithreading branch which is approved and it changes a lot of things. If we merge it in the master anytime soon, I can wait until then and rebase it. 

Please share your suggestions.

apoorvyadav1111 commented:
Hi @JyotinderSingh , and @AshwinKul28, I have addressed the comments as well as rebased with the master branch. The build and lints checks have been successful, and the new commands are per the new store logic. 

Please review and share your views. If all looks good, please merge.

Thanks,
Apoorv

JyotinderSingh commented on core/t_set.go:
```diff
@@ -0,0 +1,310 @@
+
+func evalSADD(args []string, store *Store) []byte {
+	if len(args) < 2 {
+		return Encode(errors.New("ERR wrong number of arguments for 'SADD' command"), false)
```
Please use standardized error builders which were recently introduced to return errors. Same for all other occurrences.

apoorvyadav1111 replied:
Hi @JyotinderSingh, I have updated the code with new error standards. I had to add a new error type as the existing one was not per the Redis SET data commands standard. Please review and merge if everything looks good.

JyotinderSingh reviewed (commented):
Thanks for this PR @apoorvyadav1111! The changes look good overall. Have left one comment.

apoorvyadav1111 commented:
Hi @JyotinderSingh, Addressed latest review comments.

JyotinderSingh reviewed (changes requested):
Thanks for addressing the reviews @apoorvyadav1111, could we move the eval functions to eval.go for consistency? We might clean them up in the future into smaller files but for now let's stick with the existing format.

apoorvyadav1111 commented:
Hi @JyotinderSingh, I moved all the set functions to `eval.go`. Please review and merge if everything looks good.

Thanks,
Apoorv

apoorvyadav1111 commented:
Hi @JyotinderSingh, I have rebased with the master, however one of the integration tests for QWATCH is failing. Would it be okay to merge this soon if the test is flaky? As the repo is growing swiftly, this branch must be rebased and force-pushed frequently, leading to worry it doesn't break existing features😅.

I will ensure that the next PR I raise isn't this big. 

Kind regards,
Apoorv

JyotinderSingh reviewed (approved):
Thanks for adding this feature @apoorvyadav1111. Approved.
Re-ran the CI and tests are now passing.
